File: 백준/Silver/24479. 알고리즘 수업 － 깊이 우선 탐색 1/알고리즘 수업 － 깊이 우선 탐색 1.py
import sys
sys.setrecursionlimit(10**6)  # 🔥 중요!

N, M, R = map(int, sys.stdin.readline().split())

adj = [[] for _ in range(N + 1)]

for _ in range(M):
    a, b = map(int, sys.stdin.readline().split())
    adj[a].append(b)
    adj[b].append(a)

for i in range(1, N + 1):
    adj[i].sort()

visited = [0] * (N + 1)
ans = [0] * (N + 1)
cnt = 0

def dfs(node):
    global cnt
    cnt += 1
    visited[node] = True
    ans[node] = cnt

    for neighbor in adj[node]:
        if not visited[neighbor]:
            dfs(neighbor)

dfs(R)

for i in range(1, N + 1):
    print(ans[i])

File: 백준/Silver/24479. 알고리즘 수업 － 깊이 우선 탐색 1/알고리즘 수업 － 깊이 우선 탐색 1.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<vector<int>> adj;
vector<int> ans;
int cnt = 1;

void dfs(int node) {
    ans[node] = cnt;
    for (auto neighbor : adj[node]) {
        if (ans[neighbor] == 0) {
            cnt++;
            dfs(neighbor);
        }
    }
}

int main() {
    int N, M, R;
    cin >> N >> M >> R;

    adj.resize(N + 1);
    ans.resize(N + 1, 0);

    for (int i = 0; i < M; i++) {
        int a, b;
        cin >> a >> b;
        adj[a].push_back(b);
        adj[b].push_back(a);
    }

    for (int i = 1; i <= N; i++) {
        sort(adj[i].begin(), adj[i].end());
    }

    dfs(R);

    for (int i = 1; i <= N; i++) {
        cout << ans[i] << '\n';
    }

    return 0;
}
